### PR TITLE
Clarify sctlr

### PR DESCRIPTION
**Pull request by etienne-lms**
This follows #1651 and attempts to clarify (and fix) initialization of SCTLR in 32bit mode generic entries.

- fix plat_cpu_reset_early() from scratching the generic inits of SCTLR. 
- Merge the sequences done in primary cold boot, and secondary boot (ATF or not).
- Introduce CFG_SCTLR_ENABLE_RR and update some platforms config.

### CODE REVIEW

**Comment by etienne-lms** [comment]
This P-R is not pushy against the 2.5.0. Please take time to review.
@jeffy1009, @MrVan, @jforissier, @jenswi-linaro.

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 1]
```diff
@@ -186,24 +224,12 @@ UNWIND(	.cantunwind)
...
-	orr	r0, r0, #(SCTLR_WXN | SCTLR_UWXN)
-#endif
-	write_sctlr r0
-	isb
-
 	/* Early ARM secure MP specific configuration */
 	bl	plat_cpu_reset_early
 
+	set_sctlr
+	isb
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: This `isb` could go into the `set_sctlr` macro too (same below)

Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: True but i preferred this way as it is aligned by the generic_entry_a64.S and make it clear there is a synchro.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: OK

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 2]
```diff
@@ -155,6 +155,45 @@ KEEP_INIT _start
...
+#if defined(CFG_ENABLE_SCTLR_RR)
+		orr	r0, r0, #SCTLR_RR
+#endif
+#if defined(CFG_ENABLE_SCTLR_Z)
+		/* Some ARMv7 architectures need btac flush with post synchro */
+		write_bpiall
+		dsb
+		isb
+		orr	r0, r0, #SCTLR_Z
+#endif
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: I think the program flow prediction could be safely enable after the mmu is setup (still upon `CFG_ENABLE_SCTLR_Z`). This way, `CFG_ENABLE_SCTLR_Z` could be default enabled for all ARMv7 core.
You're feeling ?

Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Sure, but it's a bit ugly to write a `res1` bit as `SCTLR_Z` on ARMv8 AArch32 so I'd prefer if we can handle that in some way.

Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: Current proposal already handles that. In **core/arch/arm/cpu/cortex-armv8-0.mk**, you found
```makefile
$(call force,CFG_ENABLE_SCTLR_RR,n)
$(call force,CFG_ENABLE_SCTLR_Z,n)
```
So ARMv8 cores will not currently never try to enable SCTLR[RR] or SCTLR[Z].

Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Thanks, forgot that. Good.

**Comment by etienne-lms** [comment]
Add fixup to enable program flow prediction enabling only once mmu is enable.

This now complies with supported ARMv7 cores: either STCLR[Z] must be enable by SW means, or if not required (recent cores), the bit is write-ignored and program flow prediction enable is tied to MMU enabling. Hence `CFG_ENABLE_SCTLR_Z` is safe to enable for all supported ARMv7 cores.

Tested on b2260 (c-a9) and qemu_virt (c-a15) with and without the branch prediction.

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 3]
```diff
@@ -155,6 +155,58 @@ KEEP_INIT _start
 	.endm
 
 	/*
+	 * set_sctlr : Setup some core configuration in CP15 SCTLR
+	 *
+	 * Followng setup is required by current core implementation:
```
Comment (jforissier) [inline core/arch/arm/kernel/generic_entry_a32.S]: s/Followng/Following/

**Comment by etienne-lms** [comment]
Rebased. Ack-by tag applied.

**Comment by jforissier** [comment]
@etienne-lms did you forget to apply Jens' R-b tags? Or did some patches change to much for you to apply them?

**Comment by etienne-lms** [comment]
Hi jerome,
There were few changes in the patches since jens' r-b. Hence i did not apply his tag, even if modifications were not that big. Feel free to wait for some r-b before merging.

**Comment by jforissier** [comment]
OK, that's fine. Let's give @jenswi-linaro some time to look at this again.

**Comment by jenswi-linaro** [comment]
As far as I can tell there's only changes in a comment since last time I looked at the code so my r-b still holds.

**Comment by etienne-lms** [comment]
Oups, bad changes pushed. Will update. Sorry.

**Comment by etienne-lms** [comment]
Much better.
Rebased. 
Added a late fix (cortex-a7 all revisions ties branch prediction and mmu control).

Will apply tags soon...

**Comment by etienne-lms** [comment]
rebased. tags applied.

**Comment by jforissier** [comment]
@etienne-lms only the first patch has Acked-by/Reviewed-by tags.

**Comment by etienne-lms** [comment]
rebased and tags applied.

**Pull request merged**